docs/testing_mbed_OS_5.md

```diff
@@ -0,0 +1,187 @@
+};
+
+Specification specification(test_setup, cases);
+
+// Entry point into the tests
+int main() {
+    return !Harness::run(specification);
+}
+```
+
+This test will first run a case that succeeds, then it will run a test that fails. This is a good template to use when creating tests. For more complex testing examples, please see the documentation for [utest](https://github.com/ARMmbed/mbed-os/tree/master/features/frameworks/utest).
+
+## Debugging tests
+
+Debugging tests is a crucial part of the development and porting process. This section will cover exporting the test, then driving the test with the test tools while the target is attached to a debugger.
+
+### Exporting tests
+
+Currently, the easiest way to export a test is to copy the test's source code from its test folder to your project's root. This way it will be treated like a normal application by the tools.
+
+You can find the path to the test you wish to export by running the following command:
+
+```
+mbed test --compile-list -n <test name>
+```
+
+Once you've copied all of the test's source files to your project root, go ahead and export your project:
+
+```
+mbed export -i <ide name>
+```
+
+Your exported project should now be in `projectfiles/<ide>_<target>`. Go ahead and open this project in your IDE.
+
+### Running a test while debugging
+
+Assuming your test has exported correctly to your IDE, go ahead and build the project and load it onto your target via your debugger.
+
+Bring the target out of reset and run the program. Your target will now be waiting for a synchronizing character string to be sent from the test tools over the serial port. The `mbed test` command cannot be run as that will attempt to flash the device, which you've already done with your IDE.
+
+Instead, the underlying test tools can be used to drive the test. [htrun](https://github.com/ARMmbed/htrun) is the tool that needs to be used in this case. This is installed when you install the requirements for mbed OS. However, if you don't have it installed you can do this by running `pip install mbed-host-tests`.
+
+First, find your target's serial port by running the following command:
+
+```
+$ mbed detect
+
+[mbed] Detected KL46Z, port COM270, mounted D:
+
+...
+```
+
+From the output, take note of where your target's serial port (in this case, it's `COM270`).
+
+Run the following command when your device is running the test in your debugger:
+
+```
+mbedhtrun --skip-flashing --skip-reset -p <serial port>:9600
+```
+
+Replace `<serial port>` with the serial port you found by running `mbed detect` above.
+
+So for the example above, the command would be:
+
+```
+mbedhtrun --skip-flashing --skip-reset -p COM270:9600
+```
+
+This will detect your attached target and drive the test. At this point the test should proceed and you should be able to debug it. If you need to rerun the test, simply flash and/or reset the device with your debugger, run the program, and run the same command.
```

Comment:
"and/or" because they may need only need to reset, not re-flash. But I agree this is clunky wording. I'll change it to just "reset".